architecture/transaction-layer.md

```diff
@@ -0,0 +1,145 @@
+
+- **Intents* for all of a transaction’s writes, which represent a provisional, uncommitted state. These are essentially the same as standard multi-version concurrency control (MVCC) values but also contain a special intent flag that points to the Transaction Record stored on the cluster. As long as the value exists as an intent, its values are treated as provisional.
+
+   As intents are written, CockroachDB checks for existing intents for the same keys. If it encounters any, it checks its transaction record; if the transaction is `PENDING`, the transaction waits for the existing one to resolve; otherwise it resolves `COMMITTED` values or removes `ABORTED` ones.
+
+If there is a deadlock, the transaction with the lowest priority has fails and returns a retryable error to the client.
+
+If transactions fail for other reasons, such as failing to pass a SQL constraint, the transaction is also aborted.
+
+### Committing Transaction (Phase 2)
+
+CockroachDB checks if the transaction has been aborted; if it has, it surfaces an error to the client. If the transaction was aborted due to a deadlock with another transaction, it will generate a retryable error, meaning that the client should restart the transaction using updated values.
+
+If the transaction is not aborted, CockroachDB then attempts to perform any read operations (e.g., `SELECT`). If the read only encounters standard MVCC values everything is fine. However, if it encounters any Write Intents, what happens next depends on the depends on the committing transaction’s isolation level. This setting is defined in the SQL statement, but is handled programmatically in the Transaction Layer.
+
+- **Serializable Snapshot Isolation** (SSI) transactions are CockroachDB's default (equivalent to ANSI SQL's `SERIALIZABLE` isolation level, which is the highest of the four standard levels). If a read encounters Write Intents from an SSI transaction, it must wait for the transaction to complete.
+
+- **Snapshot Isolation** (SI) transactions provide slightly weaker guarantees than SSI ones, allowing an anomaly called [write skew](https://en.wikipedia.org/wiki/Snapshot_isolation) to occur. In return, they perform better under certain kinds of contention. This is similar to the ANSI SQL `REPEATABLE READ` isolation level (but not exactly equivalent).
+   
+   If a read encounters an intent written by an SI transaction, it doesn't have to wait for the transaction to complete. Instead, it can "push" the writing transaction's timestamp forward, and allow the read to proceed. This improves performance of reads when read/write conflicts occur, but again, at the expense of allowing an anomaly into the data.
```

Comment:
> Instead, it can "push" the writing transaction's timestamp forward, and allow the read to proceed.

This sentence isnt clear by itself. At this point in the document we haven't explained the relationship between timestamps and ability to proceed.

Reply:
Created a new section and linked to it from this sentence. I also think that there's a somewhat intuitive sense for people who are reading the architecture docs deeply that moving timestamps in a database creates anomalies, which is the thing I'm mostly trying to communicate here.